src/Core/Security/Voter/PermissionVoter.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Forumify\Core\Security\Voter;
+
+use Forumify\Core\Entity\User;
+use Forumify\Plugin\Entity\Permission;
+use Forumify\Core\Repository\PermissionRepository;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\Voter;
+
+class PermissionVoter extends Voter
+{
+    private PermissionRepository $permissionRepository;
+
+    public function __construct(PermissionRepository $permissionRepository)
+    {
+        $this->permissionRepository = $permissionRepository;
+    }
+
+    protected function supports(string $attribute, $subject): bool
+    {
+        return preg_match('/^[a-z]+\.[a-z]+\.[a-z]+\.[a-z]+$/', $attribute) === 1;
+    }
+
+    protected function voteOnAttribute(string $attribute, $subject, TokenInterface $token): bool
+    {
+        $user = $token->getUser();
+
+        if (!$user instanceof User) {
+            return false;
+        }
+
+        $permission = $this->permissionRepository->findOneBy(['permission' => $attribute]);
```

Comment:
This will probably become like

- iterate over all roles in user
- if `in_array($permission, $role->getPermissions())` => return true